plan/logical_plan_builder.go

```diff
@@ -921,40 +929,20 @@ type ApplyConditionChecker struct {
 // innerPlan. This way is the so-called correlated execution.
 func (b *planBuilder) buildApply(outerPlan, innerPlan LogicalPlan, checker *ApplyConditionChecker) LogicalPlan {
 	ap := &Apply{
-		InnerPlan:        innerPlan,
-		Checker:          checker,
-		corColsInCurPlan: make([]*expression.CorrelatedColumn, len(outerPlan.GetSchema())),
-		baseLogicalPlan:  newBaseLogicalPlan(App, b.allocator),
+		InnerPlan:       innerPlan,
+		Checker:         checker,
+		baseLogicalPlan: newBaseLogicalPlan(App, b.allocator),
 	}
 	ap.self = ap
 	ap.initID()
 	addChild(ap, outerPlan)
-	_, innerPlan, b.err = innerPlan.PredicatePushDown(nil)
-	if b.err != nil {
-		return nil
-	}
-	corColumns, err := innerPlan.PruneColumnsAndResolveIndices(innerPlan.GetSchema())
-	if err != nil {
-		b.err = errors.Trace(err)
-		return nil
-	}
+	corColumns := innerPlan.extractCorrelatedCols()
+	ap.correlated = outerPlan.IsCorrelated()
 	for _, corCol := range corColumns {
 		// If the outer column can't be resolved from this outer schema, it should be resolved by outer schema.
 		if idx := outerPlan.GetSchema().GetIndex(&corCol.Column); idx == -1 {
-			ap.corColsInOuterPlan = append(ap.corColsInOuterPlan, corCol)
-		} else {
-			if ap.corColsInCurPlan[idx] == nil {
-				ap.corColsInCurPlan[idx] = &expression.CorrelatedColumn{
-					Column: *outerPlan.GetSchema()[idx],
-					Data:   &types.Datum{},
-				}
-			}
-			corCol.Data = ap.corColsInCurPlan[idx].Data
-		}
-	}
-	for i := len(ap.corColsInCurPlan) - 1; i >= 0; i-- {
-		if ap.corColsInCurPlan[i] == nil {
-			ap.corColsInCurPlan = append(ap.corColsInCurPlan[:i], ap.corColsInCurPlan[i+1:]...)
+			ap.correlated = true
```

Comment:
I use `false` to replace `true` that can pass all test in the package of TiDB.  It may be something wrong or need to add some tests.